csharp-tree-sitter.el

```diff
@@ -274,30 +277,51 @@
     (modify-syntax-entry ?@ "_" table)
     table))
 
-;;;###autoload
-(define-derived-mode csharp-tree-sitter-mode prog-mode "C#"
-  "Major mode for editing Csharp code.
+(defvar-local csharp-tree-sitter--syntax-table nil)
+(defvar-local csharp-tree-sitter--mode-map nil)
+(defvar-local csharp-tree-sitter--indent-line-function nil)
 
-Key bindings:
-\\{csharp-tree-sitter-mode-map}"
-  :group 'csharp
-  :syntax-table csharp-tree-sitter-mode-syntax-table
+(defun csharp-tree-sitter--enable ()
+  "Enable tree-sitter support for csharp."
+  ;; NOTE: Record lost data before configure to `tree-sitter' support.
+  (setq csharp-tree-sitter--syntax-table csharp-mode-syntax-table
+        csharp-tree-sitter--mode-map csharp-mode-map
+        csharp-tree-sitter--indent-line-function indent-line-function)
 
-  (setq-local tree-sitter-indent-current-scopes csharp-mode-indent-scopes)
-  (setq-local tree-sitter-indent-offset csharp-mode-indent-offset)
-  (setq-local indent-line-function #'tree-sitter-indent-line)
+  (setq-local csharp-mode-syntax-table csharp-tree-sitter-mode-syntax-table
+              csharp-mode-map csharp-tree-sitter-mode-map)
+  (setq-local tree-sitter-indent-current-scopes csharp-mode-indent-scopes
+              tree-sitter-indent-offset csharp-mode-indent-offset
+              indent-line-function #'tree-sitter-indent-line)
 
   ;; https://github.com/ubolonton/emacs-tree-sitter/issues/84
   (unless font-lock-defaults
     (setq font-lock-defaults '(nil)))
   (setq-local tree-sitter-hl-default-patterns csharp-mode-tree-sitter-patterns)
   ;; Comments
-  (setq-local comment-start "// ")
-  (setq-local comment-end "")
+  (setq-local comment-start "// " comment-end "")
+
+  (tree-sitter-hl-mode 1))
 
-  (tree-sitter-hl-mode))
+(defun csharp-tree-sitter--disable ()
+  "Disable tree-sitter support for csharp."
+  ;; NOTE: Revert all previous configuration from `csharp-mode'.
+  (setq csharp-mode-syntax-table csharp-tree-sitter--syntax-table
+        csharp-mode-map csharp-tree-sitter--mode-map
+        indent-line-function csharp-tree-sitter--indent-line-function)
+  (tree-sitter-hl-mode -1))
 
-(add-to-list 'tree-sitter-major-mode-language-alist '(csharp-tree-sitter-mode . c-sharp))
+;;;###autoload
+(define-minor-mode csharp-tree-sitter-mode
+  "Minor mode for csharp tree-sitter support."
+  :lighter " C#-TreeSitter"
+  :group csharp
+  (if (and (eq major-mode 'csharp-mode) csharp-tree-sitter-mode)
+      (csharp-tree-sitter--enable)
+    (csharp-tree-sitter--disable)
+    (user-error "You cannot enable tree-sitter support outside of `csharp-mode`")))
+
+(add-to-list 'tree-sitter-major-mode-language-alist '(csharp-mode . c-sharp))
```

Comment:
Just use `csharp-mode`.

Reply:
This is already included in `tree-sitter`, so we don't have to provide that. I included it in the other mode in master since that one is not included. This line can go away